This patch adds support for 'dynamic partitions'.  This basically 
works as follows:
* The nand code generates a bad-block-table at the first scan of the chip
* The dynamic partition code calculates the raw partition sizes based on 
  the bad block table.  E.g. if you have a partition of size 0x30000, and there are 
  two bad blocks (0x4000 each) in it, the raw size will increase to 0x38000, and the
  following partitions get shifted towards the end of flash.

Please note that currently the desired partition sizes are stored at compile-time
in an array in drivers/nand/nand_bbt.c, so this definitely needs to change before
submitting/merging upstream.

In order to calculate the partiton map (and set mtdparts accordingly), you can use
the 'dynpart' command at the prompt.  Use 'saveenv' to make the setting permanent.

Signed-off-by: Harald Welte <laforge@openmoko.org>

Index: u-boot.git/drivers/nand/nand_bbt.c
===================================================================
--- u-boot.git.orig/drivers/nand/nand_bbt.c	2007-02-06 12:10:58.000000000 +0100
+++ u-boot.git/drivers/nand/nand_bbt.c	2007-02-06 12:11:58.000000000 +0100
@@ -1051,4 +1051,54 @@
 	return 1;
 }
 
+#define ARRAY_SIZE(x)           (sizeof(x) / sizeof((x)[0]))
+static unsigned int part_size[] = { 0x30000, 0x4000, 0x200000, 0x400000, 0x39cc000 };
+static char *part_names[] = { "u-boot", "u-boot env", "kernel", "initrd", "rootfs" };
+
+int nand_create_mtd_dynpart(struct mtd_info *mtd)
+{
+	struct nand_chip *this = mtd->priv;
+	int part;
+	char *mtdparts;
+	unsigned int cur_offs = 0;
+
+	mtdparts = malloc(512); /* FIXME: bounds checking */
+	if (!mtdparts)
+		return -ENOMEM;
+
+	memset(mtdparts, 0, sizeof(mtdparts));
+	sprintf(mtdparts, "mtdparts=gta01-0:");
+
+	for (part = 0; part < ARRAY_SIZE(part_size); part++) {
+		unsigned int bb_delta = 0;
+		unsigned int offs = 0;
+		char mtdpart[32];
+		for (offs = cur_offs; offs < cur_offs + part_size[part]; offs += 0x4000) {
+			int block = (int) (offs >> (this->bbt_erase_shift - 1));
+			u_int8_t bbt = (this->bbt[block >> 3] >> (block & 0x06)) & 0x03;
+			if (bbt == 0x03)
+				bb_delta += 0x4000;
+		}
+
+		if (cur_offs + part_size[part] + bb_delta > this->chipsize)
+			part_size[part] = this->chipsize - cur_offs - bb_delta;
+#if 0
+		printf("partition %u: start = 0x%08x, end=%08x size=%08x, size_inc_bb=%08x\n",
+			part, cur_offs, cur_offs + part_size[part] + bb_delta,
+			part_size[part], part_size[part] + bb_delta);
+#endif
+		cur_offs += part_size[part] + bb_delta;
+		sprintf(mtdpart, "0x%.8x(%.16s),", part_size[part] + bb_delta, part_names[part]);
+		mtdpart[sizeof(mtdpart)-1] = '\0';
+		strncat(mtdparts, mtdpart, sizeof(mtdparts)-strlen(mtdparts)-1);
+	}
+
+	mtdparts[strlen(mtdparts)-1] = '\0';
+	printf("mtdparts %s\n", mtdparts);
+	setenv("mtdparts", mtdparts);
+
+	free(mtdparts);
+	return 0;
+}
+
 #endif
Index: u-boot.git/include/configs/neo1973.h
===================================================================
--- u-boot.git.orig/include/configs/neo1973.h	2007-02-06 12:11:34.000000000 +0100
+++ u-boot.git/include/configs/neo1973.h	2007-02-06 12:11:58.000000000 +0100
@@ -103,7 +103,7 @@
 			CFG_CMD_ELF	 | \
 			CFG_CMD_MISC	 | \
 			/* CFG_CMD_USB	 | */ \
-			/* CFG_CMD_JFFS2	 | */ \
+			CFG_CMD_JFFS2	 | \
 			CFG_CMD_DIAG	 | \
 			/* CFG_CMD_HWFLOW	 | */ \
 			CFG_CMD_SAVES	 | \
@@ -211,13 +211,13 @@
 #define CONFIG_FAT		1
 #define CONFIG_SUPPORT_VFAT	1
 
-#if 0
+#if 1
 /* JFFS2 driver */
 #define CONFIG_JFFS2_CMDLINE	1
 #define CONFIG_JFFS2_NAND	1
 #define CONFIG_JFFS2_NAND_DEV	0
-#define CONFIG_JFFS2_NAND_OFF	0x634000
-#define CONFIG_JFFS2_NAND_SIZE	0x39cc000
+//#define CONFIG_JFFS2_NAND_OFF	0x634000
+//#define CONFIG_JFFS2_NAND_SIZE	0x39cc000
 #endif
 
 /* ATAG configuration */
@@ -253,4 +253,7 @@
 
 #define CONFIG_S3C2410_NAND_BBT		1
 
+#define MTDIDS_DEFAULT	"nand0=gta01-0"
+#define MTPARTS_DEFAULT	"gta01-0:192k(u-boot),4k(u-boot env),2M(kernel),4M(initrd),-(jffs2)"
+
 #endif	/* __CONFIG_H */
Index: u-boot.git/common/cmd_jffs2.c
===================================================================
--- u-boot.git.orig/common/cmd_jffs2.c	2007-02-06 11:38:01.000000000 +0100
+++ u-boot.git/common/cmd_jffs2.c	2007-02-06 12:11:58.000000000 +0100
@@ -2123,6 +2123,18 @@
 	printf ("Usage:\n%s\n", cmdtp->usage);
 	return 1;
 }
+
+extern int nand_create_mtd_dynpart(struct mtd_info *mtd);
+
+int do_dynpart(cmd_tbl_t *cmdtp, int flag, int argc, char *argv[])
+{
+#if 0
+	int i = simple_strtoul(argv[1], NULL, 0);
+	if (i >= CFG_MAX_NAND_DEVICE)
+		return -EINVAL;
+#endif
+	nand_create_mtd_dynpart(&nand_info[0]);
+}
 #endif /* #ifdef CONFIG_JFFS2_CMDLINE */
 
 /***************************************************/
@@ -2188,6 +2200,13 @@
 	"<name>     := '(' NAME ')'\n"
 	"<ro-flag>  := when set to 'ro' makes partition read-only (not used, passed to kernel)\n"
 );
+
+U_BOOT_CMD(
+	dynpart, 1,	1,	do_dynpart,
+	"dynpart\t- dynamically calculate partition table based on BBT\n",
+	"\n"
+	"    - sets 'mtdparts' according to BBT\n");
+
 #endif /* #ifdef CONFIG_JFFS2_CMDLINE */
 
 /***************************************************/
